Fix issues with CI on forks

Sjors opened the pull request:
Maintainer note: `SKIP_BRANCH_PUSH=true` must be set in Cirrus for `bitcoin-core/gui` before merging this. See `https://cirrus-ci.com/github/bitcoin-core/gui` -> Settings.

---

I find myself making pull requests against my fork (mostly on top of https://github.com/bitcoin/bitcoin/pull/28983, or asking others to do so. Currently only the Github actions are run on forks, because we use self-hosted runners for the Cirrus tasks.

While setting up my own self-hosted runners for my fork, I ran into a number of issues. Some of those were addressed by https://github.com/bitcoin/bitcoin/pull/29441, but remaining issues are:

1. When PRs are opened in the fork, cirrus CI jobs are run twice because PRs and branches reside in the same repository, rather than a main repository and a fork repository, as is the case with bitcoin/bitcoin PRs. Fix this by adding a `SKIP_BRANCH_PUSH` configuration option that allows skipping CI runs not directly associated with a PR. The fix is a generalization of [#20328](https://github.com/bitcoin/bitcoin/pull/20328), which fixed a similar problem for the bitcoin-core/gui mirror repository, and it allows removing a hardcoded reference to that repository.

    Github actions jobs will still run twice despite this change, see [#29274 (comment)](https://github.com/bitcoin/bitcoin/pull/29274#issuecomment-2188840483). Initially this PR tried to prevent that with https://github.com/bitcoin/bitcoin/commit/b9fdd0dc75b5b4944dffc700b0391b38465f754a, but this had some potentially negative side effects, see [#29274 (comment)](https://github.com/bitcoin/bitcoin/pull/29274#discussion_r1457587805), so that commit was dropped for now.

2. When PRs are opened in the fork, the "test-each-commit" github action can fail due to not being able to find a recent merge commit. This problem doesn't happen in the bitcoin/bitcoin repository because branches in this repository used as the base for pull requests always point at merge commits.

This PR replaces https://github.com/bitcoin/bitcoin/pull/29259 using the self hosted workers via Cirrus instead of Github.

You can see this PR in action on this pull request to my fork: https://github.com/Sjors/bitcoin/pull/30

To test it yourself:

1. spin up at least two [self hosted runners](https://github.com/cirruslabs/cirrus-cli/blob/master/PERSISTENT-WORKERS.md). Either use a seperate VM for each, or give them their own user.
3. Install Podman and other CI dependencies (see .cirrus.yml)
4. Give Cirrus access to your fork at https://cirrus-ci.com/settings/github/YOU
5. Get a token from Cirrus and use it to start your worker(s)
6. Optionally set SKIP_BRANCH_PUSH=true ~and NO_ARM=true~ env variables (see .cirrus.yml)
make a pull request to your own fork, with this PR as the base branch

Security wise: when dealing with code from strangers on the internet, review it first before running the CI. There's a Cirrus check-box that requires approval for people without write access to trigger CI.

Sjors commented on .cirrus.yml:
```diff
@@ -161,11 +172,13 @@ task:
 
 task:
   name: 'ASan + LSan + UBSan + integer, no depends, USDT'
+  # Skip if no Noble worker exists (for a fork)
```
It stays pending, at least in the Cirrus interface.

Sjors replied:
Trying that UX in https://github.com/Sjors/bitcoin/pull/36 ...

Sjors replied:
On the first push of a branch, due to the UI glitched described below (https://github.com/bitcoin/bitcoin/pull/29274#issuecomment-1967030916) they show as skipped:

<img width="875" alt="Scherm­afbeelding 2024-02-27 om 17 40 10" src="https://github.com/bitcoin/bitcoin/assets/10217/6a76f6f1-25dc-4235-b04b-0e605da8f7fc">

That seems fine. On the second push I expect them to stay "in progress" forever. Will add a screenshot later.

I agree that hiding the task altogether is not ideal (as ecf01f42c4df4193098e1edb7a8e19c42677d2ef does). Perhaps marking it as skipped is better.

fanquake replied:
Just generally, I'd rather not add more code/complication to our CI infra to work around quirks/annoying side-effects in other infra, especially when the number of people affected by them are very close to 0.

fanquake replied:
>  it's currently quite tedious to run CI on forks,, which encourages people to make pull requests to the main repo that could be avoided.

Maybe, but anyone can currently run any of the CI jobs they'd like, locally, against any code they'd like (not sure what forks, or nested PRs has to do with it, as long as there is a git branch with our ci infra, the CI job can run?). It seems that the requirement here is to have a Cirrus/GitHub interface?

Sjors replied:
It's not that many lines really (apart from documentation), especially as @maflcko has been chipping away at the commits.

A few more lines can be dropped when Ubuntu 24.04 is out. efc22a9ae7b4d1b346e5f1ac8578927a27286917 can be dropped in a followup if the underlying issue is solved.

I don't mind giving people the option to use Microsoft, but that doesn't remove the need to support self-hosting imo.

> the requirement here is to have a Cirrus/GitHub interface?

Yes, for the same reason we have that on the main repo: it's a much lower barrier to entry. It helps the author by catching mistakes early and it gives reviewers the option to avoid wasting time on things that don't pass CI.

The use case I have in mind is someone who wants to improve the Template Provider in #29432 (before it's merged). I have dozens of TODO's in that PR which I'd like help with. In order to get the benefit of a full CI run, they currently have to open a PR to the main repo. With the tweaks in my PR they can simply open a PR to my fork instead. I think that's better for everyone.

Sjors replied:
Well for one thing it's _way_ faster.

Sjors replied:
It's also not this black and white. Sure, Microsoft and Cirrus can screw us over by faking the CI result. But by not hosting them for the machines themselves, it's both easier to catch them doing them than and easier to move elsewhere.

Sjors replied:
If I run a self-hosted CI I can see what it's doing. If it fails on my machine and is marked as successful on Github, I can see that. If I collude with Github that's a different story of course.

Sjors commented on .github/workflows/ci.yml:
```diff
@@ -9,7 +9,10 @@ on:
+      # Do not run branch pushes on forks, because they are typically redundant with
+      # pull requests to the main repo.
+      - 'bitcoin/**'
+      - 'bitcoin-core/**'
```
Good point, this probably needs an ENV flag.

Sjors replied:
Knots pull requests seem to be typically made against the `25.x-knots` branch. Those would run normally. But with the above code CI would not run when e.g. a future `26.x.knots` is created (or rebased).

Inquisition seems to follow the some pattern.

Sjors replied:
Aaargh, afaik Github CI doesn't support custom ENV variables for forks. So this is solved, but on Github forks are unconditionally skipped.

Sjors committed:
ci: forks can opt-out of CI branch push (Cirrus only)

Sjors commented on ci/test/02_run_container.sh:
```diff
@@ -52,7 +56,7 @@ if [ -z "$DANGER_RUN_CI_ON_HOST" ]; then
                   --mount "type=volume,src=${CONTAINER_NAME}_depends_SDKs_android,dst=$DEPENDS_DIR/SDKs/android" \
                   --mount "type=volume,src=${CONTAINER_NAME}_previous_releases,dst=$PREVIOUS_RELEASES_DIR" \
-                  --env-file /tmp/env \
+                  --env-file /tmp/env-$USER \
```
I don't think you're supposed to do that. One user == one worker.

If we want to support multiple workers per user, we also have to randomise the container names, since you might be running CI for two pull requests at the same time.

Sjors replied:
> Could be a separate pull request, because this is a bug fix?

I might split it out if this PR takes too long on its own.

Sjors replied:
Ok, so maybe `/tmp/env-$USER-$CONTAINER_NAME`

Sjors commented on .cirrus.yml:
```diff
@@ -45,17 +45,26 @@ env:  # Global defaults
 # - arm64: For an aarch64 machine, recommended to have 2 CPUs and 8 GB of memory.
+#
+# For works without sudo rights, start them with USER_MODE=true and make sure:
+# - git is installed
```
Oh, I assume it that explicit install was there for a reason, the comment "Unconditionally install git (used in fingerprint_script)." is a bit vague.

Sjors commented on .cirrus.yml:
```diff
@@ -45,17 +45,26 @@ env:  # Global defaults
     # Unconditionally install git (used in fingerprint_script).
-    - bash -c "$PACKAGE_MANAGER_INSTALL git"
+    # In USER_MODE this must have been done manually
+    - if [ ! "$USER_MODE" = "true" ]; then bash -c "$PACKAGE_MANAGER_INSTALL git"; fi
```
That's indeed nicer, if the line can't be dropped entirely.

Sjors commented on ci/lint/06_script.sh:
```diff
@@ -11,7 +11,7 @@ set -ex
   # To faithfully recreate CI linting locally, specify all commits on the current
   # branch.
-  COMMIT_RANGE="$(git merge-base HEAD master)..HEAD"
+  COMMIT_RANGE="$(git merge-base HEAD origin/master)..HEAD"
```
Preserving @maflcko's comment on the previous PR:

> unrelated: Generally I'd find it better if we get rid of the commit range completely. For example, when running the whitespace linter, it seems easier to just lint all code at `HEAD`, instead of trying to guess which code was modified and only lint that.

Sjors replied:
I remember this fixed an error complaining that there's no `master` branch. But I can retest.

Sjors replied:
IIUC only the cirrus linter CI uses calls this script. So it's possible this commit was only useful in #29259, where I added that linter to a Github Action. I'll drop it if CI passes on https://github.com/Sjors/bitcoin/pull/30

Sjors replied:
I dropped 0ac57f8becd10277549d1612a11b3b7c48b1fd6a.

Someone can drop `COMMIT_RANGE` altogether in a separate PR. Maybe move the incantation to a README, because it's certainly useful to know how to run the linter on each commit.

Sjors commented on .github/workflows/ci.yml:
```diff
@@ -29,7 +29,10 @@ jobs:
     runs-on: ubuntu-22.04
-    if: github.event_name == 'pull_request' && github.event.pull_request.commits != 1
+    # Only run on pull requests, skip if there's only commit.
+    # Do not run on forks.
```
It would be nicer to fix the actual problem, but I don't understand what this job is doing.

Sjors replied:
@hebasto @ryanofsky any thoughts on how to make the `test-each-commit` task work for a scenario like in https://github.com/Sjors/bitcoin/pull/30 where the base branch is `fork/some-branch` and the PR branch is `fork/some-pr`?

Sjors replied:
Example of earlier failure: https://github.com/Sjors/bitcoin/actions/runs/7556559882/job/20573828392?pr=28#step:4:42

```
 shell: /usr/bin/bash -e {0}
  env:
    DANGER_RUN_CI_ON_HOST: 1
    CI_FAILFAST_TEST_LEAVE_DANGLING: 1
    MAKEJOBS: -j10
    MAX_COUNT: 6
    FETCH_DEPTH: 8
Warning: you are leaving 1 commit behind, not connected to
any of your branches:

  3b5c5d4 ci: add self-hosted runners for GitHub

If you want to keep it by creating a new branch, this may be a good time
to do so with:

 git branch <new-branch-name> 3b5c5d4

HEAD is now at cbe740b ci: vary /tmp/env
fatal: bad revision '^^@'
```

ajtowns commented:
> I'm somewhat puzzled why Knots (cc @luke-jr) and Inquisition (cc @ajtowns) are not running into the same issues. But I guess there's at least two differences:

Inquisition disabled all the automatic tests except the trivial linter when cirrus imposed limits earlier in the year, and tests are only run manually by reviewers. Haven't worked out what it should be doing, so don't have suggestions to offer, but interested to see what the resolution here is.

luke-jr commented:
>I'm somewhat puzzled why Knots and Inquisition are not running into the same issues

Maybe because I switch the self-hosted CI to not-self-hosted?

fanquake commented on ci/test/02_run_container.sh:
```diff
@@ -30,6 +30,10 @@ if [ -z "$DANGER_RUN_CI_ON_HOST" ]; then
 
   if [ -n "${RESTART_CI_DOCKER_BEFORE_RUN}" ] ; then
+    # Kill harder, prevents "The container name "/ci_..." is already in use by container"
+    docker stop "$(docker ps -a -q)" || true
```
What is "kill harder"? Containers are already being `--all --force` removed below:
```bash
-a, --all                   Remove all containers
-f, --force                 Force removal of a running or unusable container
```
so they shouldn't need to be "stopped" beforehand?

Also, shouldn't this be `podman`?

Sjors replied:
It was intentionally vague, because I might just be doing something else wrong. I'll read @maflcko's suggestion.

Sjors replied:
Oh, this is only available in Noble? In that case I might just improve the code comment (including `sleep`) and add a note that can go away with Noble.

Sjors replied:
I installed Podman 4.8.3 from source on my machine. However I don't understand what you mean by `run --replace`. It seems we're only using podman in the `RESTART_CI_DOCKER_BEFORE_RUN`, but still call `docker run` everywhere. Can we just swap `docker run` with `podman run` or is it more complicated?

fanquake replied:
I don't have docker installed on any machine I use for CI. Only podman, and a docker -> podman alias (podman-docker).

Sjors replied:
Mmm, I tried something similar in de26cbc79230f64d7747b8b88f4dee71ebd37e1b by adding a `DOCKER_RUN` variable and setting it to `podman run --replace`. But this fails with `Error: short-name resolution enforced but cannot prompt without a TTY`. Are you passing any other arguments in your alias?

fanquake replied:
> Are you passing any other arguments in your alias?

The alias came from:
```bash
apt install podman-docker
# podman-docker is already the newest version (4.8.3+ds1-2).
docker --version
# podman version 4.8.3
podman --version
# podman version 4.8.3
```

Sjors replied:
podman-docker is not documented in the manual installation... https://podman.io/docs/installation#building-from-source

Let's see how I can install this without totally messing up my OS :-)

Sjors replied:
That base64 decodes to:

```sh
#!/usr/bin/env bash

echoerr() { echo "$@" 1>&2; }

echoerr " (:::docker->)podman ${@}"

podman "${@}"
```

That doesn't seem any different from my `DOCKER_RUN="podman run --replace"` alias.

Sjors replied:
Or I guess it's the `echoerr` doing some magic

Sjors replied:
Yes, Docker is installed in rootless mode: https://docs.docker.com/engine/security/rootless/

Forgot to link to the error: https://cirrus-ci.com/task/5247017476161536?logs=ci#L261

The "short-name resolution" error makes no sense to me based on what I can find about it, since the container URL includes `docker.io` (`docker.io/amd64/ubuntu:22.04` in the above example).

> you can't install podman-docker without removing docker

Ok, let's not go that route, since Docker currently works.

> because the cleanup is done with the podman command, which can not clear docker containers

Ah, so that explains why I needed 7cdb75d572e825b4c0cf74a516f912f94807853f.

> This should fix the issue you are trying to "fix" in this commit.

Why do we add a dependency on podman at all? Given that `docker stop "$(docker ps -a -q)" || true` works. I guess it's less clean because it cleans all docker processes run by the user. But is there some other benefit? Is the plan to move to podman entirely?

Sjors replied:
Oh, I see where my confusion comes from. I thought podman was just a wrapper for Docker. But it's actually an alternative which can run the same containers. Why not switch over entirely to Podman? It's quite confusing that we're using both.

Sjors replied:
I nuked Docker in favor of Podman. So far it seems to work. I dropped 7cdb75d572e825b4c0cf74a516f912f94807853f.

willcl-ark replied:
Not specifically related to the issue at hand: the new podman desktop app is pretty smart, it can show you containers running via docker and podman (and kubernetes clusters) all in the same UI: https://podman-desktop.io/

fanquake commented on ci/test/02_run_container.sh:
```diff
@@ -30,6 +30,10 @@ if [ -z "$DANGER_RUN_CI_ON_HOST" ]; then
   if [ -n "${RESTART_CI_DOCKER_BEFORE_RUN}" ] ; then
+    # Kill harder, prevents "The container name "/ci_..." is already in use by container"
+    docker stop "$(docker ps -a -q)" || true
+    sleep 3
```
Also, why is this `sleep` here?

Sjors replied:
Without that command I got the same behavior as if the command wasn't there. Presumably it's asynchronous and we somehow have to wait for it. Maybe the above solution avoids the problem altogether.

Sjors replied:
I got rid of sleep, in favor of `docker rm -f $CONTAINER_NAME` (haven't pushed that change yet, pending discussion on podman)

https://stackoverflow.com/questions/57631654/how-to-wait-for-a-docker-container-to-be-removed

Sjors replied:
These lines are dropped now.

Sjors commented:
I switched my own machine from Docker to Podman. This removes the need for 7cdb75d572e825b4c0cf74a516f912f94807853f so that's dropped.

Added a commit that explains the process a bit more clearly.

Sjors commented:
Rebased after #29441 landed which contains two commits from this PR.

Sjors commented:
Dropped 0ac57f8becd10277549d1612a11b3b7c48b1fd6a since that wasn't doing anything. Also rebased.

Sjors commented:
One annoying side-effect of 80f338c01623469bd3347d8a2e438294512c141c is that if you open a pull request (after pushing the branch) the Cirrus tasks initially all show as skipped. The tasks are actually running though, and as they finish that's reflected on Github.

This only happens for Cirrus workers, not for Github workers (3048700bde41644f1b8b40825334b6ecc0f7ccc1).

See screenshots: https://github.com/Sjors/bitcoin/pull/36#issuecomment-1966883011

This seems like a bug in Github or in Cirrus's integration. Dropping 80f338c01623469bd3347d8a2e438294512c141c has two undesirable effects:
1. It runs CI twice for every push
2. It runs my self-hosted CI for every push to pull requests on the main repo

The new behaviour is opt-in by setting `NO_BRANCH=true`, so everyone can decide what they prefer. But perhaps there's a more elegant solution here?

Sjors commented:
I pushed two changes:

1. 80f338c01623469bd3347d8a2e438294512c141c -> 0d084293e3083ffda6309cb80c06852b3010f869: use `only_if:` instead of `skip:`: this slightly improves the experience of when you open a new PR.

2. ecf01f42c4df4193098e1edb7a8e19c42677d2ef -> 7f7f87b6691e222b9488e55b2a41a2e763071ead: unsupported architecture are marked skipped instead of hidden (except for the first time you open a PR, which hopefully is fixed with https://github.com/cirruslabs/cirrus-ci-docs/issues/1233)

So all pushes except the initial one look like this:
<img width="529" alt="Scherm­afbeelding 2024-02-27 om 19 09 37" src="https://github.com/bitcoin/bitcoin/assets/10217/f1d4bf01-9d9f-41d7-9978-4262b3cbf495">

luke-jr commented on .cirrus.yml:
```diff
@@ -8,18 +8,35 @@ env:  # Global defaults
+# Generally, a persistent worker must run Ubuntu 23.04+ or Debian 12+.
+#
+# The following specific types should exist, with the following requirements:
+# - small: For an x86_64 machine, recommended to have 2 CPUs and 8 GB of memory.
+# - medium: For an x86_64 machine, recommended to have 4 CPUs and 16 GB of memory.
+# - noble: For a machine running the Linux kernel shipped with exaclty Ubuntu Noble 24.04. The machine is recommended to have 4 CPUs and 16 GB of memory.
+# - arm64: For an aarch64 machine, recommended to have 2 CPUs and 8 GB of memory.
```
```suggestion
# - small: For an x86_64 machine, recommended to have 2 CPU cores and 8 GB of memory.
# - medium: For an x86_64 machine, recommended to have 4 CPU cores and 16 GB of memory.
# - noble: For a machine running the Linux kernel shipped with exactly Ubuntu Noble 24.04. The machine is recommended to have 4 CPU cores and 16 GB of memory.
# - arm64: For an aarch64 machine, recommended to have 2 CPU cores and 8 GB of memory.
```

Sjors replied:
Done

luke-jr reviewed (commented):
This seems more like "*don't* support Cirrus on forks", since it disables it to a large extent. Maybe that should also be behind env vars?

Sjors commented:
> Maybe that should also be behind env vars?

It is. Nothing is disabled for Cirrus on forks, unless you set `NO_BRANCH`, `NO_ARM` and/or `NO_NOBLE`.

On Github CI there's no way (afaik) to set env vars, so there a few things are disabled for forks.

luke-jr commented on .cirrus.yml:
```diff
@@ -8,24 +8,41 @@ env:  # Global defaults
   CCACHE_NOHASHDIR: "1"  # Debug info might contain a stale path if the build dir changes, but this is fine
 
+# A self-hosted machine(s) can be used via Cirrus CI. It can be configured with a
+# single user with sudo access, or multiple users without by setting USER_MODE=true
```
What acts on this env var?

luke-jr commented on .cirrus.yml:
```diff
@@ -8,24 +8,41 @@ env:  # Global defaults
+# - podman-docker-4.1+ is required due to the bugfix in 4.1
 #   (https://github.com/bitcoin/bitcoin/pull/21652#issuecomment-1657098200)
-# - The ./ci/ depedencies (with cirrus-cli) should be installed:
+# - The ./ci/ depedencies (with cirrus-cli) should be installed. One-liner example
```
dependencies*

Sjors commented on .cirrus.yml:
````diff
@@ -8,24 +8,41 @@ env:  # Global defaults
 #   ```
 #
-# - There are no strict requirements on the hardware, because having less CPUs
+# - There are no strict requirements on the hardware, because having less CPU cores
````
Changed to "having fewer CPU threads run".

Also rebased.

Sjors commented:
Rebased. I dropped `NO_NOBLE` since Noble has been released. Dropped 8b06d6fb797b1de36abb13f3aaf1a7d1607c2685 after #29660.

Sjors commented:
I had to increase the fetch depth in 6398efbe145fb297957060600615262d59f2b47c, because otherwise the test-each-commit task can fail with: `fatal: bad revision '^^@'` and `You are not currently on a branch`, see [example](https://github.com/Sjors/bitcoin/actions/runs/9267073776/job/25492701755?pr=30#step:4:42).

This happened in https://github.com/Sjors/bitcoin/pull/30 because it doesn't build on a merge commit but rather on top of the branch in this PR. I'm surprised it doesn't also fail on a PR like https://github.com/bitcoin/bitcoin/pull/28201/commits.

ryanofsky commented on .github/workflows/ci.yml:
```diff
@@ -35,7 +35,9 @@ jobs:
-        run: echo "FETCH_DEPTH=$((${{ github.event.pull_request.commits }} + 2))" >> "$GITHUB_ENV"
+        # Fetch some extra commits for pull requests on forks that don't build
+        # on a merge commit.
+        run: echo "FETCH_DEPTH=$((${{ github.event.pull_request.commits }} + 100))" >> "$GITHUB_ENV"
```
In commit "ci: increase fetch depth for 'test each commit'" (6398efbe145fb297957060600615262d59f2b47c)

I think the following would be a more direct fix. I didn't test it but it just omits the `^${MERGE_BASE}^@` argument when there is no merge base, which should be fine because there should be no need to exclude merge base ancestors in that case:

```diff
-          echo "TEST_BASE=$(git rev-list -n$((${{ env.MAX_COUNT }} + 1)) --reverse HEAD ^$(git rev-list -n1 --merges HEAD)^@ | head -1)" >> "$GITHUB_ENV"
+          MERGE_BASE=$(git rev-list -n1 --merges HEAD)
+          EXCLUDE_MERGE_BASE_ANCESTORS=
+          if test -n "$MERGE_BASE"; then
+              EXCLUDE_MERGE_BASE_ANCESTORS=^${MERGE_BASE}^@
+          fi
+          echo "TEST_BASE=$(git rev-list -n$((${{ env.MAX_COUNT }} + 1)) --reverse HEAD $EXCLUDE_MERGE_BASE_ANCESTORS | head -1)" >> "$GITHUB_ENV"
```

Sjors replied:
Thanks, will try.

ryanofsky commented on .github/workflows/ci.yml:
```diff
@@ -62,7 +62,12 @@ jobs:
-          echo "TEST_BASE=$(git rev-list -n$((${{ env.MAX_COUNT }} + 1)) --reverse HEAD ^$(git rev-list -n1 --merges HEAD)^@ | head -1)" >> "$GITHUB_ENV"
+          MERGE_BASE=$(git rev-list -n1 --merges HEAD)
+          EXCLUDE_MERGE_BASE_ANCESTORS=
+          if test -n "$MERGE_BASE"; then
```
In commit "ci: test-each-commit merge base optional" (65abc0b5b54edb06cc3bf584691beddbeb802ddc)

Maybe add comment like "MERGE_BASE can be empty due to limited fetch-depth". Otherwise I don't think it's clear why git wouldn't be able to find the merge commit.

ryanofsky commented on .cirrus.yml:
```diff
@@ -48,7 +48,13 @@ env:  # Global defaults
+  # https://cirrus-ci.org/guide/writing-tasks/#conditional-task-execution
+  skip: $CIRRUS_REPO_FULL_NAME == "bitcoin-core/gui" && $CIRRUS_PR == ""
+  # Allow forks to skip CI runs when a branch is pushed.
+  # Use only_if rather than skip, because otherwise all jobs are marked as
+  # skipped when a pull request is first opened.
```
In commit "ci: forks can opt-out of CI branch push" (8fbf7449af130126890932986a4d9c0446c2f9c8)

I don't really understand this comment. Maybe it could say why the jobs should not be marked as skipped, or maybe say how they should be shown instead of how they should not be shown.

Sjors replied:
Improved the comment, see below.

ryanofsky commented on .cirrus.yml:
```diff
@@ -48,7 +48,13 @@ env:  # Global defaults
+  # Allow forks to skip CI runs when a branch is pushed.
+  # Use only_if rather than skip, because otherwise all jobs are marked as
+  # skipped when a pull request is first opened.
+  only_if: $NO_BRANCH != "true" || $CIRRUS_PR != ""
```
In commit "ci: forks can opt-out of CI branch push" (8fbf7449af130126890932986a4d9c0446c2f9c8)

What is $NO_BRANCH? Is it a custom variable like $NO_ARM added later? Could it be documented like NO_ARM is?

Sjors replied:
It's indeed a custom variable like `NO_ARM`. I moved the explanation of `NO_BRANCH` up in the documentation. Leaving only the explanation of `skip` vs. `only_if` here, which I also reworded.

ryanofsky commented on .github/workflows/ci.yml:
```diff
@@ -9,7 +9,10 @@ on:
   push:
     branches:
-      - '**'
+      # Do not run branch pushes on forks, because they are typically redundant with
+      # pull requests to the main repo.
```
In commit "ci: skip Github CI on branch pushes for forks" (1a06d80e7bb113b9588efc113262bdd20b60efd4)

I didn't really understand this comment until I read the PR description. Might be clearer as "Disable CI on branch pushes to forks. It will still run for pull requests. This prevents CI from running twice for typical pull request workflows."

Also, I'm not really sure how this change relates to the rest of the PR. It seems like it could be unrelated?

Sjors replied:
Updated.

It serves the same purpose as `NO_BRANCH` on Cirrus, expect that I couldn't find a way to make that behaviour opt-in on Github CI.

Sjors replied:
I moved 063ea4e0da9e1c06cc0f4aec97a03abac66ae59f to be right after 8259a13954878fd6188bb218c3b7320385516596 and expanded the commit message to explain why they don't use the same mechanism.

ryanofsky reviewed (approved):
Code review ACK 65abc0b5b54edb06cc3bf584691beddbeb802ddc with caveat that I don't know enough about cirrus and github configuration to be very confident in the changes. But they do make sense and appear to do what's described.

Sjors commented:
Rebased after #30193, which removes "noble" as a machine type. I previously already removed `NO_NOBLE`, see https://github.com/bitcoin/bitcoin/pull/29274#issuecomment-2134764402. The rebase merely impacts the documented list of persistent workers, which this PR moves.

Addressed review comments.

Sjors commented on .cirrus.yml:
```diff
@@ -115,7 +115,9 @@ task:
   name: 'ARM, unit tests, no functional tests'
+  # Skip if no ARM worker exists (for a fork)
   << : *GLOBAL_TASK_TEMPLATE
+  skip: $NO_ARM == "true"
```
Is that new behavior? Let me test that again...

Sjors replied:
I guess I never tested this, because the documentation just says:

```
# - arm64: For an aarch64 machine, recommended to have 2 CPUs and 8 GB of memory.
```

If the performance difference isn't too terrible, then I can just drop the commit and mention this in the documentation.

Sjors replied:
`qemu-aarch64` adds quite a few packages to Ubuntu, so I'd probably want to keep the `NO_ARM` option around, but document this alternative.

Sjors replied:
I guess this involves partially reverting #28087 since it seems that stripped the `qemu-arm` call from `00_setup_env_arm.sh`.

Sjors replied:
(comments came in out of order for me)

I don't understand Docker, Podman and/or Qemu well enough to figure this out myself. For now I'll just leave this commit.

Sjors replied:
I'm fine with just skipping that action on the stratum v2 based fork.

It's probably not too much hassle if there's clear instructions on how to do that on e.g. an AMD based Ubuntu machine.

The other way around doesn't work for me, because I don't have a native ARM machine (mac or otherwise). I'll eventually have one, a laptop. I might then use that for just the ARM task, not the rest of CI.

> moving it to GHA

I don't know how common ARM is for servers these days? I'm assuming it's mostly used by Apple, and for very low performance devices like Raspberry Pi's.

Sjors replied:
After `podman run` is there some VM running that I have to login to, install the Cirrus job runner? Or is there some change needed to `00_setup_env_arm.sh` to make it use this?

Sjors replied:
Thanks. I currently use `podman-docker` on Ubuntu 24.04. I created several `cirrus{1,2,3,4,5}` users.

For `cirrus1` I set `type: arm64`. I turned off `NO_ARM=true` for the occasion.

I installed `qemu-user-static` (`bash` and `podman-docker` were already there).

I ran your `podman run ...` incantation as the `cirrus1` user. And then started the worker:

```
RESTART_CI_DOCKER_BEFORE_RUN=1  USER_MODE=true MAKEJOBS=-j10 cirrus worker run -f ~/.cirrus.yaml
```

I then force-pushed an update to https://github.com/Sjors/bitcoin/pull/30. The ARM task fails: https://cirrus-ci.com/task/6683160417665024

```
STEP 6/7: RUN ["bash", "-c", "cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh"]
exec container process `/usr/bin/bash`: Exec format error
Error: building at STEP "RUN bash -c cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh": while running runtime: exit status 1
```

Sjors replied:
The CPU is an AMD Ryzen 9 7950X.

Sjors replied:
That first command:

```
WARNING: image platform (linux/arm64/v8) does not match the expected platform (linux/amd64)
{"msg":"exec container process `/usr/bin/uname`: Exec format error","level":"error","time":"2024-06-20T09:03:31.220776Z"}
```

Running the second command does not change anything.

Sjors replied:
```
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 9 7950X 16-Core Processor
Virtualization:                       AMD-V
VERSION="24.04 LTS (Noble Numbat)"
podman version 4.9.3
```

Sjors replied:
```
podman run --rm --privileged docker.io/multiarch/qemu-user-static --reset -p yes
Setting /usr/bin/qemu-alpha-static as binfmt interpreter for alpha
Setting /usr/bin/qemu-arm-static as binfmt interpreter for arm
Setting /usr/bin/qemu-armeb-static as binfmt interpreter for armeb
Setting /usr/bin/qemu-sparc-static as binfmt interpreter for sparc
Setting /usr/bin/qemu-sparc32plus-static as binfmt interpreter for sparc32plus
Setting /usr/bin/qemu-sparc64-static as binfmt interpreter for sparc64
Setting /usr/bin/qemu-ppc-static as binfmt interpreter for ppc
Setting /usr/bin/qemu-ppc64-static as binfmt interpreter for ppc64
Setting /usr/bin/qemu-ppc64le-static as binfmt interpreter for ppc64le
Setting /usr/bin/qemu-m68k-static as binfmt interpreter for m68k
Setting /usr/bin/qemu-mips-static as binfmt interpreter for mips
Setting /usr/bin/qemu-mipsel-static as binfmt interpreter for mipsel
Setting /usr/bin/qemu-mipsn32-static as binfmt interpreter for mipsn32
Setting /usr/bin/qemu-mipsn32el-static as binfmt interpreter for mipsn32el
Setting /usr/bin/qemu-mips64-static as binfmt interpreter for mips64
Setting /usr/bin/qemu-mips64el-static as binfmt interpreter for mips64el
Setting /usr/bin/qemu-sh4-static as binfmt interpreter for sh4
Setting /usr/bin/qemu-sh4eb-static as binfmt interpreter for sh4eb
Setting /usr/bin/qemu-s390x-static as binfmt interpreter for s390x
Setting /usr/bin/qemu-aarch64-static as binfmt interpreter for aarch64
Setting /usr/bin/qemu-aarch64_be-static as binfmt interpreter for aarch64_be
Setting /usr/bin/qemu-hppa-static as binfmt interpreter for hppa
Setting /usr/bin/qemu-riscv32-static as binfmt interpreter for riscv32
Setting /usr/bin/qemu-riscv64-static as binfmt interpreter for riscv64
Setting /usr/bin/qemu-xtensa-static as binfmt interpreter for xtensa
Setting /usr/bin/qemu-xtensaeb-static as binfmt interpreter for xtensaeb
Setting /usr/bin/qemu-microblaze-static as binfmt interpreter for microblaze
Setting /usr/bin/qemu-microblazeel-static as binfmt interpreter for microblazeel
Setting /usr/bin/qemu-or1k-static as binfmt interpreter for or1k
Setting /usr/bin/qemu-hexagon-static as binfmt interpreter for hexagon
```

Sjors replied:
When I run both commands as root, it does work. But doing so does not change anything for the non-root user. At least that's a clue.

Sjors replied:
Some background information on what this actually does (though not podman specific): https://stackoverflow.com/a/72890225

Sjors replied:
Adding `cirrus1` to `/etc/sudoers` and running both commands with `sudo` works. But running only the `--reset` command with `sudo` is not enough.

Not really a solution, but at least I can run the CI job once to see what the performance is like.

Sjors replied:
Uhhh ... so I did a reboot. And now `podman run --rm -ti "docker.io/arm64v8/debian:bookworm" uname --machine` just works, doesn't need the reset command.

Sjors replied:
If the job finishes in reasonable time, I'll add something like this to the documentation: https://github.com/Sjors/bitcoin/commit/d065dc11fb862f48fc88e674a95b7daf68269ca0

It seems that this either works trivially or is really hard to debug, so I prefer to keep `NO_ARM=1` around as an alternative.

Sjors replied:
1 hour 42 minutes, but that was without using any cache.

Sjors replied:
58 minutes for the second run, which is acceptable.

Sjors replied:
Without this change, when there's no ARM machine, it remains pending forever.

fanquake replied:
> I understand. However, my point is that this is a cosmetic thing. Who cares? The person that sets up a fork, enables cirrus on it, enables a worker pool on it, knowingly does not enable an arm worker on it, so the person should know that the arm task will not run. Whatever the icon in the UI displays should thus not matter.

I agree with Marco. We seem to be spending a lot of time, and happy to take on some complexity just to accommodate use-cases *outside* this repository; for which, as far as I can see, there is not actually that much demand. We also shouldn't have to introduce complexity here, just to fix cosmetic issues in other repositories.

To me it's also unclear what the requirements are here going forward. I assume if any of the outside-of-this-repo functionality breaks, eventually someone might notice, and a fix might happen, but at the same time, I don't think anyone in this repo is going to be thinking about this when making code changes here (and they shouldn't have to), so things may break, and that will be unfortunate, but should never be blocking changes here. We are maintaining our infrastructure; not our infrastructure plus the infrastructure of whoever happened to fork this repository.

Sjors replied:
It's confusing for both the pull request author and reviewers. Only the fork maintainer will know what's going on, but they also need to carefully check each time that it's only the expected job that's stuck. A single green check, when all jobs pass, is more clear.

> does yaml allow duplicate keys

Not sure.

So the simplest solution for now is to drop 859e1c558fd9c8604fda361e465aa4eb4a676823 and make qemu "mandatory", which works for me.

> just to accommodate use-cases outside this repository

But it reduces the number of direct pull requests _to_ repository, so on balance it may save time.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29274#discussion_r1652884524

I don't have a strong opinion, but I do think the NO_ARM commit 859e1c558fd9c8604fda361e465aa4eb4a676823 is a nice change because I don't think it fixes a purely cosmetic issue. I think it removes confusing behavior of opening a PR and seeing a job hang forever. The actual code change is very simple, just adding `skip: $NO_ARM == "true"` and I think the documentation change makes the whole "A self-hosted machine(s) can be used via Cirrus CI..." section clearer and more concrete. But I agree with dropping it as long as any objections remain.

Sjors replied:
It's conceptually simple, but I agree that we shouldn't use duplicate keys unless Cirrus docs explicitly says you can. I could work around it in various ways, but for now I found it easier to just drop the commit.

Sjors commented on .cirrus.yml:
```diff
@@ -8,25 +8,45 @@ env:  # Global defaults
-# - There are no strict requirements on the hardware, because having less CPUs
-#   runs the same CI script (maybe slower). To avoid rare and intermittent OOM
+# - There are no strict requirements on the hardware, because having fewer CPU threads
+#   run the same CI script (maybe slower). To avoid rare and intermittent OOM
```
Changed to:

```cpp
# - There are no strict requirements on the hardware. Having fewer CPU threads
#   than recommended merely causes the CI script to run slower.
```

Sjors replied:
The list of machines only contain recommendations for CPUs and memory. Presumably giving them less memory than recommended can lead to OOM. Giving them fewer CPUs is the only recommendation you can safely ignore (if you're not in a CI environment that kills jobs after a timeout).

Sjors commented on .cirrus.yml:
```diff
@@ -8,25 +8,45 @@ env:  # Global defaults
 #
-# Generally, a persistent worker must run Ubuntu 23.04+ or Debian 12+.
-# Specifically,
+# When a contributor makes a pull request against a fork of the repo, two CI
```
Oh, so only the maintainer of the fork will have this problem? Maybe there's another solution then.

Sjors replied:
Ah wait, now I remember, reworded:

```
# When a contributor maintains a fork of the repo, any pull request they make
# to their own fork, or to the main repository, will trigger two CI runs:
# one for the branch push and one for the pull request.
# This can be avoided by setting NO_BRANCH=1 as a custom env variable in Cirrus.
```

Sjors commented:
I turns out it's fairly straight-forward to run the arm native job on an x86_64 machine using `qemu-user-static` (ht @maflcko). I added a note on how to do that. But I'm keeping 247b5823b2d5810ac9ddb8c8562232ff3eeb0344 around as an alternative.

Sjors commented:
@maflcko done in #30314, minus the paragraphs about `NO_ARM` and `NO_BRANCH`.

Sjors commented on .cirrus.yml:
```diff
@@ -44,10 +44,21 @@ env:  # Global defaults
+  # No need to run on the read-only mirror, unless it is a PR.
+  # https://cirrus-ci.org/guide/writing-tasks/#conditional-task-execution
+  skip: $CIRRUS_REPO_FULL_NAME == "bitcoin-core/gui" && $CIRRUS_PR == ""
+  # We use only_if rather than skip when NO_BRANCH is set. Otherwise when a
+  # contributor opens a pull request, they will initially see all jobs marked
+  # as skipped.
```
What I remember from a while ago, but I haven't retested, is that under a pull request all the checks initially are marked as skipped. But then they show as either finished or failed as the results come in. It was rather flaky.

I could try again and make some screenshots...

Sjors replied:
Glad I checked this, because the glitched described here is no longer happening:
https://github.com/bitcoin/bitcoin/pull/29274#issuecomment-1967030916

ryanofsky replied:
> I could try again and make some screenshots...

Thanks, I wouldn't make an extra effort to understand the broken behavior you were trying to avoid. But maybe consider  dropping the word "initially" because I think the comment would make more sense without it. The idea of not avoiding "skip" directive to avoid tests showing up as "skipped" is fairly clear. It's just not clear why they tests might initially show up as skipped but then later show up not skipped, if that's what happens.

It might also help if this said what the code does, instead of just what it does not do. The line
`only_if: $NO_BRANCH != "true" || $CIRRUS_PR != ""` is a hard to parse, but I think it means: Run this job if is part of a PR, and also run it on standalone branches, as the project has not set a custom NO_BRANCH variable. Not too important, but I think it could help if comment said before describing why skip was not used.

Sjors commented:
Simplified c02d96f8f5829d433f4a061480aa37ee0c6ff076 because the UI glitch I noticed by in February no longer happens. See https://github.com/bitcoin/bitcoin/pull/29274#discussion_r1647869054

Rebased on #30314, so will temporarily mark this PR as draft.

ryanofsky commented on .github/workflows/ci.yml:
```diff
@@ -61,7 +61,13 @@ jobs:
           # ancestors from the rev-list output as described in:
           # https://git-scm.com/docs/git-rev-list
-          echo "TEST_BASE=$(git rev-list -n$((${{ env.MAX_COUNT }} + 1)) --reverse HEAD ^$(git rev-list -n1 --merges HEAD)^@ | head -1)" >> "$GITHUB_ENV"
+          MERGE_BASE=$(git rev-list -n1 --merges HEAD)
```
In commit "ci: test-each-commit merge base optional" (b80a573bb42e10f1ec550c79133317c1c0b021c6)

Might be useful to expand commit message to explain why this change is needed. Maybe say:

- The ci "test-each-commit" job fetches the PR branch being tested with a depth of (# of commits in PR + 2), and then tries to run tests on commits after the most recent merge commit.

   When a PR is opened against a bitcoin core branch, a merge commit is always guaranteed to exist within the fetch depth, because bitcoin core branches always point at merge commits.

  However, in fork repositories, pull requests can be opened based on other branches that don't contain recent merge commits, and this will currently cause git rev-list to fail with `fatal: bad revision '^^@'`.

  Work around this problem by not requiring a recent merge commit, and just testing on all fetched commits if a merge commit can't be found.

ryanofsky reviewed (approved):
Code review ACK b80a573bb42e10f1ec550c79133317c1c0b021c6, but I would think about dropping the second commit ("ci: skip Github CI on branch pushes for forks" (b9fdd0dc75b5b4944dffc700b0391b38465f754a)) and moving it to another PR, since it is a little confusing and maybe could use more discussion.

Otherwise looks very good, I just left some suggestions to try to improve some comments.

Sjors commented on .cirrus.yml:
```diff
@@ -59,7 +64,11 @@ env:  # Global defaults
 # https://cirrus-ci.org/guide/tips-and-tricks/#sharing-configuration-between-tasks
 filter_template: &FILTER_TEMPLATE
-  skip: $CIRRUS_REPO_FULL_NAME == "bitcoin-core/gui" && $CIRRUS_PR == ""  # No need to run on the read-only mirror, unless it is a PR. https://cirrus-ci.org/guide/writing-tasks/#conditional-task-execution
+  # No need to run on the read-only mirror, unless it is a PR.
+  # https://cirrus-ci.org/guide/writing-tasks/#conditional-task-execution
+  skip: $CIRRUS_REPO_FULL_NAME == "bitcoin-core/gui" && $CIRRUS_PR == ""
```
Yes, I could even change that here if someone sets `NO_BRANCH=true` before merging this.

Sjors replied:
Done

Sjors committed:
ci: test-each-commit merge base optional

The ci "test-each-commit" job fetches the PR branch being tested with a depth of (# of commits in PR + 2), and then tries to run tests on commits after the most recent merge commit.

When a PR is opened against a bitcoin core branch, a merge commit is always guaranteed to exist within the fetch depth, because bitcoin core branches always point at merge commits.

However, in fork repositories, pull requests can be opened based on other branches that don't contain recent merge commits, and this will currently cause git rev-list to fail with fatal: bad revision '^^@'.

Work around this problem by not requiring a recent merge commit, and just testing on all fetched commits if a merge commit can't be found.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

Sjors commented:
I dropped b9fdd0dc75b5b4944dffc700b0391b38465f754a as suggested by @ryanofsky so the PR is focussed on Cirrus.

I updated 026d86427527a6c58425896066386713b9782275 to drop the `bitcoin-core/gui` workaround. Updated the PR description to make clear that repo needs to have `NO_BRANCH=true` set _before_ merging.

https://github.com/bitcoin/bitcoin/pull/29274#discussion_r1652116935

Sjors commented on .cirrus.yml:
```diff
@@ -24,6 +24,9 @@ env:  # Global defaults
 # podman or docker. Background: https://stackoverflow.com/a/72890225/313633
 #
+# Alternatively arm tasks can be skipped by setting NO_ARM=1 as a custom env
+# variable in Cirrus.
```
Fixed `NO_ARM=true`

Joined the paragraphs.

Explained what happens when you do neither.

Sjors replied:
But this commit is dropped now.

Sjors commented on .cirrus.yml:
```diff
@@ -27,6 +27,11 @@ env:  # Global defaults
+# When a contributor maintains a fork of the repo, any pull request they make
+# to their own fork, or to the main repository, will trigger two CI runs:
+# one for the branch push and one for the pull request.
+# This can be avoided by setting NO_BRANCH=1 as a custom env variable in Cirrus.
```
Fixed.

I added an instruction at the top of this PR for where to set `NO_BRANCH=true`.

I can't find a Cirrus documentation page to point to for this, but you'll find it once you find the repo settings page.

Sjors commented on .cirrus.yml:
```diff
@@ -59,7 +64,11 @@ env:  # Global defaults
+  # https://cirrus-ci.org/guide/writing-tasks/#conditional-task-execution
+  skip: $CIRRUS_REPO_FULL_NAME == "bitcoin-core/gui" && $CIRRUS_PR == ""
+  # Allow forks to skip CI runs when a branch is pushed.
+  skip: $NO_BRANCH == "true" && $CIRRUS_PR == ""
```
Done

Sjors commented:
Dropped the `NO_ARM` commit as well, see https://github.com/bitcoin/bitcoin/pull/29274#discussion_r1652953447

fanquake commented:
> Maintainer note: NO_BRANCH=true must be set in Cirrus for bitcoin-core/gui before merging this. See https://cirrus-ci.com/github/bitcoin-core/gui -> Settings.

Can you rename this to something that actually explains what it's doing; `NO_BRANCH` is incredibly vauge. Why not `DONT_RUN_CI_ON_BRANCH_PUSH_ON_THIS_FORK` etc? Then someone looking at the settings in Cirrus will actually know what this is doing.

Sjors commented:
Renamed `NO_BRANCH` to `SKIP_BRANCH_PUSH`.

ryanofsky commented:
> Maintainer note: `SKIP_BRANCH_PUSH=true` must be set in Cirrus for `bitcoin-core/gui` before merging this. See `https://cirrus-ci.com/github/bitcoin-core/gui` -> Settings.

I just now went to that URL and set this. There were not any other environment variables set.

I wonder if there is a good place to document this configuration, but probably it's not too important as the main effect is just to avoiding wastefully running CI twice on pushes to master (at least according to the original PR which added the skip statement: https://github.com/bitcoin/bitcoin/pull/20328)

ryanofsky reviewed (approved):
Code review ACK 0882babb102108a4351714cc3749fceba1ecbb35.

---

I think the current PR title / description are a little hard to understand. Maybe consider editing it to something like the following:

- Fix issues with CI on forks

  I find myself making pull requests against my fork (mostly on top of [#28983](https://github.com/bitcoin/bitcoin/pull/28983), or asking others to do so. Currently only the Github actions are run on forks, because we use self-hosted runners for the Cirrus tasks.

  While setting up my own self-hosted runners for my fork, I ran into a number of issues. Some of those were addressed by [#29441](https://github.com/bitcoin/bitcoin/pull/29441), but remaining issues are:

  1. When PRs are opened in the fork, cirrus CI jobs are run twice because PRs and branches reside in the same repository, rather than a main repository and a fork repository, as is the case with bitcoin/bitcoin PRs. Fix this by adding a `SKIP_BRANCH_PUSH` configuration option that allows skipping CI runs not directly associated with a PR. The fix is a generalization of [#20328](https://github.com/bitcoin/bitcoin/pull/20328), which fixed a similar problem for the bitcoin-core/gui mirror repository, and it allows removing a hardcoded reference that repository.

      Github actions jobs will still run twice despite this change, see [#29274 (comment)](https://github.com/bitcoin/bitcoin/pull/29274#issuecomment-2188840483). Initially this PR tried to prevent that with b9fdd0dc75b5b4944dffc700b0391b38465f754a, but this had some potentially negative side effects, see [#29274 (comment)](https://github.com/bitcoin/bitcoin/pull/29274#discussion_r1457587805), so that commit was dropped for now.

  2. When PRs are opened in the fork, the "test-each-commit" github action can fail due to not being able to find a recent merge commit. This problem doesn't happen in the bitcoin/bitcoin repository because branches in this repository used as the base for pull requests always point at merge commits.

This PR replaces #29259 using the self hosted workers via Cirrus instead of Github.

You can see this PR in action on this pull request to my fork: https://github.com/Sjors/bitcoin/pull/30

To test it yourself:
1. spin up at least two [self hosted runners](https://github.com/cirruslabs/cirrus-cli/blob/master/PERSISTENT-WORKERS.md). Either use a seperate VM for each, or give them their own user.
2. Install Podman and other CI dependencies (see `.cirrus.yml`)
3. Give Cirrus access to your fork at `https://cirrus-ci.com/settings/github/YOU`
4. Get a token from Cirrus and use it to start your worker(s)
5. Optionally set `NO_BRANCH=true` ~and `NO_ARM=true`~ env variables (see `.cirrus.yml`)
6. make a pull request to your own fork, with this PR as the base branch

Security wise: when dealing with code from strangers on the internet, review it first before running the CI. There's a Cirrus check-box that requires approval for people without write access to trigger CI.

Sjors commented on .cirrus.yml:
```diff
@@ -27,6 +27,11 @@ env:  # Global defaults
+# When a contributor maintains a fork of the repo, any pull request they make
+# to their own fork, or to the main repository, will trigger two CI runs:
+# one for the branch push and one for the pull request.
+# This can be avoided by setting SKIP_BRANCH_PUSH=true as a custom env variable in Cirrus.
```
I find this equally confusing, but was a bit reluctant to add a manual for Cirrus. Anyway, added.

Sjors commented:
@ryanofsky thanks for the PR description rewrite!

ryanofsky reviewed (approved):
Code review ACK 576828e732bacb61b608cd89f669a2936d3e8d00.

I also tweaked formatting in the PR description to fix a few problems in the markdown copy / paste.

The suggested text "Fix issues with CI on forks" in the description was actually meant to be a replacement title for the PR, since the second commit in the PR fixes an issue with a github actions job, not a cirrus job. So could remove this and maybe update the title.

maflcko reviewed (approved):
lgtm ACK

My preference would be to have the `.cirrus.yml` as single source of ground truth. However, requiring forks to update this file is too cumbersome, so this alternative seems fine.

(The "config resolution strategy" is another setting outside the yaml that needs to be modified to "merge for prs", which could be documented in a follow-up)

But this looks good in any case.

Sjors commented:
@maflcko fixed in description

ryanofsky commented:
@maflcko is your "lgtm ACK" more of a partial ACK, or do you think this is good to merge? I think practically speaking this looks ok to merge, but I wouldn't want to merge it myself if I'm the only full reviewer.

I guess there are also number of possible followups to this PR if it is merged:

- Maybe preventing github actions from running twice on PRs, [#29274 (comment)](https://github.com/bitcoin/bitcoin/pull/29274#issuecomment-2188840483).
- Maybe restoring ability to skip cirrus ARM jobs [#29274 (comment)](https://github.com/bitcoin/bitcoin/pull/29274#discussion_r1653288456).
- Maybe documenting need to set bitcoin-core/gui `SKIP_BRANCH_PUSH` value somewhere [#29274 (comment)](https://github.com/bitcoin/bitcoin/pull/29274#issuecomment-2189439362).
- Maybe documenting need to set "config resolution strategy" somewhere [#29274 (comment)](https://github.com/bitcoin/bitcoin/pull/29274#pullrequestreview-2140849311).

Sjors commented:
Picked @ryanofsky's first suggestion into #30487.

So far ARM jobs have not been a source of headaches on my own fork (only TSAN and MSAN are atm).
